### PR TITLE
Bug: Fix `coosort`

### LINKED ISSUE #2005: bug in sparse matrix conversion COO->CSR for some matrices
I am finding that conversion of COO->CSR or CSC does not agree with scipy when the underlying matrix is large (nnz > 50k elements or so). The problem does not seem to occur for small matrices such as those used in the test suite.

* Conditions (you can just paste the output of `python -c 'import cupy; cupy.show_config()'`)

CuPy Version          : 6.0.0b2
CUDA Root             : /usr/local/cuda
CUDA Build Version    : 10000
CUDA Driver Version   : 10000
CUDA Runtime Version  : 10000

* Code to reproduce

The data for the example below is ~1MB in size and is available via:
https://drive.google.com/open?id=1BRaTNQoAYJPOjfGyio51OwwmCZdh3vdT

```python
import scipy
import numpy as np
import cupy
import cupyx

data = np.load('/tmp/coo_to_csc_example_100k.npz')

sl = slice(None)
A_cpu = scipy.sparse.coo_matrix((data['data'][sl],
                                 (data['row'][sl], data['col'][sl])))

A_gpu = cupyx.scipy.sparse.coo_matrix(A_cpu)

# this comparison is okay
cupy.testing.assert_allclose(A_gpu.data, A_cpu.data)

# conversion on the GPU does not match for large matrices
A_csr_gpu = A_gpu.tocsr()
A_csr_cpu = A_cpu.tocsr()
cupy.testing.assert_allclose(A_csr_gpu.data,
                             A_csr_cpu.data)

# convert via round-trip to CPU works fine
A_csr_gpu2 = cupyx.scipy.sparse.coo_matrix(A_gpu.get().tocsr())
cupy.testing.assert_allclose(A_csr_gpu2.data,
                             A_csr_cpu.data)
```

And oddly, the data array after conversion does match at the beginning and then has huge mismatches in many of the later elements.  I plot this below:

```python

import matplotlib.pyplot as plt
plt.figure()
plt.plot(A_csc_cpu.data - A_csc_gpu.data.get())
```
figure_12

The input data has values in the range [0, 1.0]. In this example the data is dtype np.float32, but the same thing occurs for float64 as well.

### LINKED ISSUE #2365: possible bug in cupyx sparse matrix transpose/multiplication
Hello cupy(x) developers,

I would like to report what looks like a bug in the cupyx sparse matrix library. I am running on a Skylake cpu and a Tesla V100 gpu on a SUSE-ish Linux system. CuPy version: 6.0.0. CUDA version 10.1.168. I am running inside a conda environment which you can find in my requirements.txt file: 

[requirements.txt](https://github.com/cupy/cupy/files/3474978/requirements.txt)

Here is a reproducer in which I am trying to compare the same operations on the cpu and gpu and checking at every step that they are the same. Finally in the last step the `np.allclose()` check fails. The `iCov_max_diff` value is much larger than machine precision which suggests there is a bug. 

```
import numpy as np
import scipy.sparse
import cupy as cp
import cupyx as cpx

m = n = 1000

np.random.seed(1)

#create random data to use in W
random_data = np.random.random(m)

#create A random sparse matrix
A_cpu = scipy.sparse.random(m, n, format='csr', random_state=42)
A_gpu = cpx.scipy.sparse.csr_matrix(A_cpu)
#yank gpu back and compare
A_yank = A_gpu.get()
assert np.allclose(A_cpu.todense(),A_yank.todense())

#create W random sparse matrix
W_cpu = scipy.sparse.spdiags(data=random_data, diags=[0,], m=m, n=n)
W_gpu = cpx.scipy.sparse.spdiags(data=random_data, diags=[0,], m=m, n=n)
#yank gpu back and compare
W_yank = W_gpu.get()
assert np.allclose(W_cpu.todense(),W_yank.todense())

#see how the dot products go
W_dot_A_cpu = W_cpu.dot(A_cpu)
W_dot_A_gpu = W_gpu.dot(A_gpu)
#yank gpu back and compare
W_dot_A_yank = W_dot_A_gpu.get()
assert np.allclose(W_dot_A_cpu.todense(), W_dot_A_yank.todense())

#check the transpose
A_trans_cpu = A_cpu.T
A_trans_gpu = A_gpu.T
#yank gpu back and compare
A_trans_yank = A_trans_gpu.get() #use get bc its a sparse object
assert np.allclose(A_trans_cpu.todense(), A_trans_yank.todense())

#okay now inverse covariance (where things go wrong)
iCov_cpu = A_cpu.T.dot(W_dot_A_cpu)
iCov_gpu = A_gpu.T.dot(W_dot_A_gpu)
#yank gpu back and compare
iCov_yank = iCov_gpu.get()

iCov_diff = iCov_cpu.todense() - iCov_yank.todense()
max_iCov_diff = np.max(iCov_diff)
print("max iCov diff")
print(max_iCov_diff)

assert np.allclose(iCov_cpu.todense(), iCov_yank.todense()) #fails for large matrix sizes
```

If there is more information I can provide please let me know. Of course if I am doing something wrong I would also be happy for your feedback about how to do this correctly. 

Thank you very much for your help,
Laurie

### PR DESCRIPTION
**Pull request by econtal**
Fixes #2005
Fixes #2365

Fix `coosort` by having `gthr` write to separated buffer instead of reading and writing to `x.data`

### CODE REVIEW

**Comment by jakirkham** [comment]
@takagi, have you had a chance to take a look at this fix? 🙂

**Comment by takagi** [comment]
pfnCI, test this please.

**Comment by takagi** [comment]
Yes, please let me check if the change works on my local as well as running the CI.

**Comment by takagi** [comment]
@econtal Sorry for leaving this PR so long. Would you fix the CI failure? It says to import `scipy` in test_cusparse.py.

**Comment by econtal** [comment]
Hi @takagi. No worries, thanks for reviewing.

I'm not familiar with cupy testing, I guess the issue was the missing `@testing.with_requires('scipy')` that I just added.

For this (and eventual further changes), do you prefer amend the commits and force push to clean the git history?

**Comment by takagi** [comment]
Thanks, you're perfect. And you don't need to clean the git history, it is a reasonable change I think.

**Comment by takagi** [comment]
pfnCI, test this please.

**Comment by takagi** [comment]
I have an issue. With this change, `x.col` and `x.row` are processed in-pace by `cusparse.xoosortByRow()`, while `x.val` is not processed in-place to be replaced by another `ndarray`. `coosort()` looks, I think, to work as an in-place procedure and it would be expected that `x.val` is sorted in-place as well as `x.col` and `x.val`.

How about sorting `x.val` in-place, for example:

```Python
data_orig = x.data.copy()
_call_cusparse(
    'gthr', x.dtype,
    handle, nnz, data_orig.data.ptr, x.data.data.ptr,
    P.data.ptr, cusparse.CUSPARSE_INDEX_BASE_ZERO)
```

It may have some time penalty compared to your change, but I suppose the consistency of memory usage is more important here. In nature, this problem comes from the different representations of a sparse matrix between cuSparse and CuPy. Three raw arrays for the former, and a wrapped-up class for the latter.

**Comment by econtal** [comment]
I see your point and agree that having hybrid in-place/copy doesn't seem like a good idea.

Copying the data to force in-place on `x.data` seems like a good way to not compromise performances too much. I'll do this change soon.

The other option would be to copy `x.col` and `x.row` and do nothing in-place, but this would be much worse impact on performances.

**Comment by econtal** [comment]
While pushing the changes I noticed that `csrsort` and `cscsort` also have this problematic pattern of reading and writing to the same buffer at the same time in `gthr`. See https://github.com/cupy/cupy/blob/master/cupy/cusparse.py#L534 and https://github.com/cupy/cupy/blob/master/cupy/cusparse.py#L562.
Also interestingly I wasn't able to create a test that would fail for `csr` or `csc`.
Do you know if there is something particular with `csr` and `csc` making it safe to call `gthr` that way? If not I'm happy to apply the fix there as well.

**Comment by takagi** [comment]
pfnCI, test this please.

**Comment by takagi** [comment]
Thanks for the change, I kicked the CI again.

As for `csrsort` and `cscsort`, I know they have the same in-place `gthr` but didn't find they work safely. I'll look into it.

**Comment by takagi** [comment]
I got a test of `csrsort` that fails. Would you try the following?

```Python
@testing.with_requires('scipy')
class TestCsrsort(unittest.TestCase):

    def setUp(self):
        self.a = scipy.sparse.random(
            1, 10000, density=0.9, dtype=numpy.float32, format='csr')
        numpy.random.shuffle(self.a.indices)
        self.a.has_sorted_indices = False

    def test_csrsort(self):
        a = sparse.csr_matrix(self.a)
        cupy.cusparse.csrsort(a)

        self.a.sort_indices()
        testing.assert_array_equal(self.a.indptr, a.indptr)
        testing.assert_array_equal(self.a.indices, a.indices)
        testing.assert_array_almost_equal(self.a.data, a.data)
```

File: cupy/cusparse.py [hunk 1]
```diff
@@ -564,6 +564,12 @@ def cscsort(x):
 
 
 def coosort(x):
+    """Sorts indices of COO-matrix in place
```
Comment (takagi) [inline cupy/cusparse.py]:
```suggestion
    """Sorts indices of COO-matrix in place.
```

**Comment by econtal** [comment]
@takagi thanks for the useful test case. I added the fix on `csrsort` and `cscsort` as well, and added corresponding tests following your suggestion.

**Comment by takagi** [comment]
pfnCI, test this please.

**Comment by takagi** [comment]
LGTM, thanks!

**Pull request merged**